contracts/StreamManager.vy

```diff
@@ -1,259 +1,539 @@
+    or service, so the security properties of this contract reflect that. As such, in all extra-
+    ordinary situations, the `controller` should have the ability or right to create and enforce
+    the terms that the payment is for, this contract simply streamlines the creation and payment of
+    those defined goods or services.
 """
-
 from ethereum.ercs import IERC20
 
 from . import Validator
 
 
+# List of Validator contracts to check for Stream creation, funding, migrations
+# TODO: Use `Set` to ensure uniqueness, when available
 MAX_VALIDATORS: constant(uint8) = 10
 validators: public(DynArray[Validator, MAX_VALIDATORS])
 
-
-MAX_REASON_SIZE: constant(uint16) = 1024
 MIN_STREAM_LIFE: public(immutable(uint256))
-
+MAX_PRODUCTS: constant(uint8) = 20
 
 struct Stream:
+    owner: address
     token: IERC20
-    amount_per_second: uint256
-    max_stream_life: uint256
     funded_amount: uint256
-    start_time: uint256
-    last_pull: uint256
-    reason: Bytes[MAX_REASON_SIZE]
-
-num_streams: public(HashMap[address, uint256])
-streams: public(HashMap[address, HashMap[uint256, Stream]])
+    expires_at: uint256
+    last_update: uint256
+    last_claim: uint256
+    products: DynArray[bytes32, MAX_PRODUCTS]
 
 
-owner: public(address)
 token_is_accepted: public(HashMap[IERC20, bool])
 
+# Global index of Streams
+num_streams: public(uint256)
+streams: public(HashMap[uint256, Stream])
+
+# Service Provider (has all Capabilities, also beneficiary of funding)
+controller: public(address)
+new_controller: public(address)
+new_controller_proposed: public(uint256)
+CONTROLLER_ACCEPTANCE_DELAY: constant(uint256) = 7 * 24 * 60 * 60  # 1 week
+
+event NewControllerProposed:
+    old_controller: indexed(address)
+    new_controller: indexed(address)
+
+
+event NewControllerAccepted:
+    old_controller: indexed(address)
+    new_controller: indexed(address)
+
+
+# Delegated Abilities
+flag Ability:
+    MODFIY_TOKENS
+    MODFIY_VALIDATORS
+    MODFIY_ACCESS
+    CANCEL_STREAMS
+
+capabilities: public(HashMap[address, Ability])
+
 
 event StreamCreated:
+    stream_id: indexed(uint256)
+    owner: indexed(address)
     token: indexed(IERC20)
-    creator: indexed(address)
+    funded_amount: uint256
+    time_left: uint256
+    products: DynArray[bytes32, MAX_PRODUCTS]
+
+
+event StreamOwnershipUpdated:
     stream_id: indexed(uint256)
-    amount_per_second: uint256
-    start_time: uint256
-    reason: Bytes[MAX_REASON_SIZE]
+    old_owner: indexed(address)
+    new_owner: indexed(address)
 
 
 event StreamFunded:
-    creator: indexed(address)
     stream_id: indexed(uint256)
-    amount_added: uint256
+    funder: indexed(address)
+    funded_amount: uint256
+    time_left: uint256
 
 
-event StreamCancelled:
-    creator: indexed(address)
+event StreamClaimed:
     stream_id: indexed(uint256)
-    amount_locked: uint256
-    reason: Bytes[MAX_REASON_SIZE]
+    claimer: indexed(address)
+    is_expires: indexed(bool)
```

Comment:
issue: I'm assuming this should be `is_expired`?

Reply:
```suggestion
    is_expired: indexed(bool)
```